cc/testing/evaluation/algorithms_with_insufficient_noise.cc

```diff
@@ -0,0 +1,187 @@
+
+    auto algorithm = absl::make_unique<CountWithInsufficientNoise<int64_t>>(
+      std::log(3), ratio);
+    
+    StochasticTester<int64_t> tester( std::move(algorithm), std::move(sequence),
+      num_datasets, num_samples_per_histogram);
+
+    bool algo_is_dp = tester.Run(); // false = 0, true = 1
+    std::cout << algo_is_dp << std::endl; 
+    return algo_is_dp; // correct outcome is false, e.g., DP premise rejected
+  }
+
+  else if (algorithm == "boundedsum") {
+
+    auto algorithm = absl::make_unique<SumWithInsufficientNoise<double>>(
+      std::log(3), sequence->RangeMin(), sequence->RangeMax(),
+      absl::make_unique<test_utils::SeededLaplaceMechanism::Builder>(), ratio);
+
+    StochasticTester<double> tester(std::move(algorithm), std::move(sequence),
+      num_datasets, num_samples_per_histogram);
+
+    bool algo_is_dp = tester.Run(); // false = 0, true = 1
+    std::cout << algo_is_dp << std::endl; 
+    return algo_is_dp; // correct outcome is false, e.g., DP premise rejected
+
+  }
+
+  else if (algorithm == "boundedmean") {
+
+    auto algorithm = absl::make_unique<MeanWithInsufficientNoise<double>>(
+      std::log(3),sequence->RangeMin(), sequence->RangeMax(),
+      absl::make_unique<test_utils::SeededLaplaceMechanism::Builder>(), ratio);
+
+    StochasticTester<double> tester(std::move(algorithm), std::move(sequence),
+      num_datasets, num_samples_per_histogram);
+
+    bool algo_is_dp = tester.Run(); // false = 0, true = 1
+    std::cout << algo_is_dp << std::endl; 
+    return algo_is_dp; // correct outcome is false, e.g., DP premise rejected
+  }
+
+  else {
+    std::cout << "The algorithm specified was not recognized." << 
+      "Try count, boundedsum, or boundedmean instead." << std::endl;
+  }
+}
+
+
+void GetTestResults(std::ofstream& datafile, std::string algorithm,
+  double num_datasets, double num_samples_per_histogram, double ratio_min,
+  double ratio_max) {
+
+  double num_tests = 0;
+  double num_tests_passed = 0;
+  double maximum_ratio_passed = 0;
+  auto start = std::chrono::high_resolution_clock::now();
+
+  for (double i=ratio_min; i<=ratio_max; i++) {
+    auto start_test_run = std::chrono::high_resolution_clock::now();
+    double ratio = i/100.0;
+    std::cout << "Now calculating " << algorithm << " algorithm with ratio: " << 
+      ratio << std::endl;
+    bool outcome = RunStochasticTester(algorithm, ratio, num_datasets,
+      num_samples_per_histogram);
+    num_tests++;
+
+    if (outcome == 0) { // correct outcome is false, e.g., DP predicate rejected 
+      num_tests_passed++;
+      maximum_ratio_passed = ratio;
+    }
+
+    auto finish_test_run = std::chrono::high_resolution_clock::now();
+    std::chrono::duration<double> test_run_length =
+      std::chrono::duration_cast<std::chrono::duration<double>>(
+      finish_test_run - start_test_run);
+    datafile << "insufficient_noise," << algorithm << ",0," << outcome << "," <<
```

Comment:
It's not clear why we always output "insufficient_noise" for the first column in the CSV. Could the first column have any other value?

In general, it would be very useful to describe somewhere what kind of output is written to the file, what values could be expected, and what those values mean.

Reply:
Yes, ideally in the final version, the file will also include other types of algorithms beyond "insufficient_noise." Specifically, the other types of unit tests will also be incorporated. 
Re: describing the output, where should I do that?

Reply:
Update: I added to the README